File: src/main/java/com/monocampusconnect/service/postgres/FacultyWorkExperienceService.java
package com.monocampusconnect.service.postgres;

import com.monocampusconnect.exception.ApiException;
import com.monocampusconnect.model.postgres.FacultyWorkExperience;
import com.monocampusconnect.repository.postgres.FacultyWorkExperienceRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class FacultyWorkExperienceService {

    private final FacultyWorkExperienceRepository facultyWorkExperienceRepository;

    public FacultyWorkExperienceService(FacultyWorkExperienceRepository facultyWorkExperienceRepository) {
        this.facultyWorkExperienceRepository = facultyWorkExperienceRepository;
    }

    public FacultyWorkExperience create(FacultyWorkExperience experience) {
        return facultyWorkExperienceRepository.save(experience);
    }

    public List<FacultyWorkExperience> getAll() {
        return facultyWorkExperienceRepository.findAll();
    }

    public FacultyWorkExperience getById(Long id) {
        return facultyWorkExperienceRepository.findById(id)
                .orElseThrow(() -> new ApiException("Faculty work experience not found", 404));
    }

    public FacultyWorkExperience update(Long id, FacultyWorkExperience experience) {
        FacultyWorkExperience existing = getById(id);
        if (experience.getFacultyId() != null) existing.setFacultyId(experience.getFacultyId());
        if (experience.getOrganization() != null) existing.setOrganization(experience.getOrganization());
        if (experience.getRole() != null) existing.setRole(experience.getRole());
        if (experience.getFromDate() != null) existing.setFromDate(experience.getFromDate());
        if (experience.getToDate() != null) existing.setToDate(experience.getToDate());
        if (experience.getDescription() != null) existing.setDescription(experience.getDescription());
        return facultyWorkExperienceRepository.save(existing);
    }

    public void delete(Long id) {
        facultyWorkExperienceRepository.delete(getById(id));
    }

    public List<FacultyWorkExperience> getByFacultyId(Long facultyId) {
        return facultyWorkExperienceRepository.findAll().stream()
                .filter(fe -> facultyId.equals(fe.getFacultyId()))
                .toList();
    }
}

File: frontend/src/lib/services/facultyWorkExperienceService.ts
import api from "../api";

interface FacultyWorkExperience {
  id?: number;
  facultyId: number;
  organization: string;
  role: string;
  fromDate?: string;
  toDate?: string;
  description?: string;
}

const ENDPOINT = "/faculty-work-experience";

export const facultyWorkExperienceService = {
  create: (data: FacultyWorkExperience) =>
    api.post<FacultyWorkExperience>(ENDPOINT, data),
  getAll: () => api.get<FacultyWorkExperience[]>(ENDPOINT),
  getById: (id: number) => api.get<FacultyWorkExperience>(`${ENDPOINT}/${id}`),
  getByFacultyId: (facultyId: number) =>
    api.get<FacultyWorkExperience[]>(`${ENDPOINT}/faculty/${facultyId}`),
  update: (id: number, data: FacultyWorkExperience) =>
    api.put<FacultyWorkExperience>(`${ENDPOINT}/${id}`, data),
  delete: (id: number) => api.delete<void>(`${ENDPOINT}/${id}`),
};

